lua/starfall/editor/docs.lua

```diff
@@ -161,54 +237,81 @@ local function parse(parsing, data)
 	end
 end
 
-local function scan(src, realm)
-	local linetbl = string.Explode("\r?\n", src, true)
-	local i = 0
-	local function lines() i=i+1 return linetbl[i] end
+local function get_lines(self)
+    local result = {}
+    local from, line_n = 1, 0
+    local delim_from, delim_to = string_find( self, "\r?\n", from  )
+    while delim_from do
+      local str = string_sub( self, from , delim_from-1 )
+      from = delim_to + 1
+      line_n = line_n + 1
+      result[line_n] = str
+      delim_from, delim_to = string_find( self, "\r?\n", from  )
+    end
+    result[line_n+1] = string_sub( self, from  )
+    return result
+end
+--- Scan function
+-- @param string src Source code
+-- @param string file_name Source file name.
+-- @param boolean type_only Whether to only parse types / classes.
+local function scan(src, realm, type_only)
+	-- https://github.com/thegrb93/StarfallEx/blob/master/lua/starfall/...
+	local file_path = string_match(curfile, "%.%./lua/starfall/(libs_.+/.*)") -- libs_sh/... path that will be used for links with [src] on the sfhelper to the github.
 	local parsing
-	for line in lines do
+
+	for line_n, line in next, get_lines(src) do
 		if parsing then
-			local data = string.match(line, "^%s*%-%-%-*(.*)")
+			local data = string_match(line, "^%s*%-%-%-*(.*)")
 			if data then
-				parse(parsing, data)
+				parse(parsing, data, type_only, line_n)
 			else
-				while line and not string.find(line, "%S") do -- Find next non-empty line
-					line = lines()
-				end
-				process(parsing, line or "")
+				process(parsing, line or "", type_only, line_n)
 				parsing = nil
 				if line == nil then break end
 			end
 		end
 		if not parsing then
-			local desc = string.match(line, "^%s*%-%-%-(.*)")
+			local desc = string_match(line, "^%s*%-%-%-(.*)")
 			if desc then
 				parsing = {
 					description = {desc},
-					realm = realm
+					realm = realm,
+					path = file_path.."#L"..line_n
 				}
 			end
 		end
 	end
 end
 
 local function realm(filename)
-	if string.match(filename, ".*_sv/[^/]+$") then
+	if string_match(filename, ".*_sv/[^/]+$") then
 		return "server"
-	elseif string.match(filename, ".*_sh/[^/]+$") then
+	elseif string_match(filename, ".*_sh/[^/]+$") then
 		return "shared"
-	elseif string.match(filename, ".*_cl/[^/]+$") then
+	elseif string_match(filename, ".*_cl/[^/]+$") then
 		return "client"
 	else
 		error("Couldn't figure out the realm! " .. filename)
 	end
 end
 
+
+-- First, go over all of the types so we have them in Docs.Types ready for typed docs.
+for name, mod in pairs(SF.Modules) do
+	for filename, data in pairs(mod) do
+		curfile = filename
+		scan(data.source, realm(filename), true)
+	end
+end
+
+-- Then, make the rest of the docs.
 for name, mod in pairs(SF.Modules) do
 	for filename, data in pairs(mod) do
 		curfile = filename
-		scan(data.source, realm(filename))
+		scan(data.source, realm(filename), false)
 		data.source = nil
 	end
 end
+
```

Comment:
Instead of scanning all the docs twice, it would be better to store the types on the first pass and then post process the data.